Addresses issue #10846

thehogfather opened the pull request:
@abose @peterflynn (apologies it isn't solely related to #10846 - it felt sensible to combine into one PR)
Fold states are now persisted using viewState (state.json)
Added copyright text into foldcode.js and foldgutter.js
Removed strings.js file
Removed latex-fold helper
Removed menu item for custom region folding
Disable custom region folding by default

thehogfather committed:
Addresses issue #10846 - fold states are now persisted using viewState (state.json)
Added copyright text into foldcode.js and foldgutter.js
Removed strings.js file
Removed latex-fold helper
Removed menu item for custom region folding
Disable custom region folding by default

abose commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -23,7 +22,7 @@ define(function (require, exports, module) {
     prefs.definePreference("alwaysUseIndentFold", "boolean", true,
                            {name: strings.ALWAYS_USE_INDENT_FOLD, description: strings.ALWAYS_USE_INDENT_FOLD_HELP});
-    prefs.definePreference("enableRegionFolding", "boolean", true,
+    prefs.definePreference("enableRegionFolding", "boolean", false,
```
If we do not expose the region fold feature by default, should we include it in the view menu?

abose replied:
Ok got it. The custom region folds feature is being removed. In that case, could't we remove the preference entry altogether, along with the menu entry and the corresponding strings?

abose commented:
Code folding is enabled in all files by default, and in some files like txt files it looks a bit odd.
Would it be a good thing to enable code folding in only HTML,XML,js Etc. and disabled in all other file types?

thehogfather commented:
@abose I suspect the fold marks are appearing in a txt file due to indentations. A suggestion would be to disable the `alwaysUseIndentFold` option so that only files with registered mode helpers for code-folding have fold markers?

thehogfather committed:
Removed string for collapse custom region menu.
Set alwaysUseIndentFold preference to be false by default
Addresses #10848 - changes to preference files are now updated live in brackets without needing to restart

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -348,5 +422,28 @@ define(function (require, exports, module) {
+                    if (editor && editor._codeMirror && CodeMirror.commands.unfoldAll) {
+                        CodeMirror.commands.unfoldAll(editor._codeMirror);
+                        foldGutter.remove(editor._codeMirror);
+                    }
```
Wouldn't you need to do this for all existing Editor instances, not just the current one?

thehogfather replied:
Yes we would. Currently the other editors open 'catch up' when the active editor changes. I couldn't find an api for iterating through open editors in the working set. Is there one? How would you do it?

peterflynn replied:
It will definitely look bad for split view, where two editors are visible at once even though only one is active.

The cleanest way to do it is walking the `_instances` array in Editor.js, which is currently inaccessible -- that's how core Editor preferences are updated.  It seems fair to me to add a public API like `forEveryEditor()` or something -- pass a callback, and it gets invoked like a forEach.

peterflynn replied:
It might be easier for me to take care of this issue in my PR -- that ok with you?

peterflynn commented on src/extensions/default/CodeFolding/foldhelpers/foldgutter.js:
```diff
@@ -310,7 +312,15 @@ define(function (require, exports, module) {
+        gutters.splice(index, 1);
+        cm.setOption("gutters",  gutters);
+        CodeMirror.defineOption("foldGutter", false, null);
+    }
```
Since this is basically the reverse of `createGutter()` in main.js, it seems like the two functions should live in the same file.  Could you either move that one here or move this there?

abose commented on src/nls/fi/strings.js:
```diff
@@ -631,7 +631,6 @@ define({
     "COLLAPSE_CURRENT"              : "Pienennä nykyinen",
     "EXPAND_CURRENT"                : "Laajenna nykyinen",
-    "COLLAPSE_CUSTOM_REGIONS"       : "Pienennä mukautetut alueet",
     "MIN_FOLD_SIZE"                 : "Pienin laskostuksen koko",
```
Only strings in the UI are put in nls/lang/strings.js file; other strings could be moved into the extension as some constants. The problem with having string constants here is that the automation team would send these strings to for translation or the community would translate these constants, which would be a loss of resources.

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -54,7 +45,11 @@ define(function (require, exports, module) {
+        collapseKey             = "Ctrl-Alt-[",
+        expandKey               = "Ctrl-Alt-]",
+        collapseAllKey          = "Alt-1",
+        expandAllKey            = "Shift-Alt-1";
```
On master, I can't seem to override any of these shortcuts using keymap.json.  I can't remember if this is some limitation of the user keybindings feature (e.g. can't override shortcuts defined by an extension? -- cc @RaymondLim)...  But to users it will just feel like a bug since they don't know which things are implemented under the hood as a built-in extension.

If it's tricky to fix we should spin this off as a separate bug though.

peterflynn replied:
Ack, scratch that -- I had the syntax backwards.  It _does_ work on master.  We should just make sure it still works in this PR too with the new, more dynamic way bindings are registered & unregistered.

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -284,10 +268,118 @@ define(function (require, exports, module) {
+        }
+
+        //create menus
+        if (!menuExists(codeFoldingMenuDivider)) {
```
I'm not clear on why we need to check all these individually.  If one of them exists, they all should...  But also, it seems like we could just maintain an internal flag like `isInitialized` to prevent doing _everything_ in `init()` when it's already been done.  (I'm assuming the core problem here is that preferences broadcasts change events sometimes when the pref value hasn't actually changed, which could lead to trying to initialize twice and thus duplicating all the menu items).

thehogfather replied:
@peterflynn I couldn't get it working properly for some reason without this smelly explicit conditions. But please feel free to have a go and update it.

peterflynn replied:
Looks like the divider gets recreated every time you set the `"enabled"` pref to true at runtime.  That's kind of an edge case -- I'll see if I can address it in my PR instead of worrying about it here.

peterflynn replied:
This part is fixed in my PR now

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -21,9 +20,9 @@ define(function (require, exports, module) {
+    prefs.definePreference("alwaysUseIndentFold", "boolean", false,
                            {name: strings.ALWAYS_USE_INDENT_FOLD, description: strings.ALWAYS_USE_INDENT_FOLD_HELP});
-    prefs.definePreference("enableRegionFolding", "boolean", true,
+    prefs.definePreference("enableRegionFolding", "boolean", false,
```
It seems like the 'custom regions' stuff is partially removed, but not fully -- the menu item and the command behind it is gone, but this preference (and its strings) is still there along with region-fold.js.  Seems like we should remove those too (and then whatever extension re-exposes the menu item could reintroduce those elements too).  Does that sound ok?

peterflynn commented:
@thehogfather This looks great overall, thanks for posting this update so quickly!  I added a few comments above.

One other note: this fixes bug #10848 for the `"enabled"` preference, but the other preferences still seem to need a restart or close/reopen file to take effect.  Did you want to wait for another PR to fix those other prefs?

peterflynn commented:
@thehogfather Fyi, I've opened PR #10859 with some UI changes as well as a few code tweaks.  Some of it duplicates change you've made here (I started working on it before you posted this), so once your PR is merged I'll rebase mine on top of your changes.  But please take a look and see if there's anything else there you'd be concerned about.

abose commented:
@thehogfather Now that region fold support is removed, fold helper - region-fold.js could also be removed.

thehogfather committed:
Moved preference strings from nls/lang to vars in Prefs.js
Changes to all preferences are now live
Removed region-fold and custom-region fold strings

thehogfather commented:
@abose @peterflynn I think this should be closer to ready than before. Cleaned up the strings and removed custom-region folding completely. Now over to you. Happy to help with anything else you direct my way.

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -8,27 +8,35 @@
+        FADE_FOLD_BUTTONS           = "Fade fold buttons",
+        FADE_FOLD_BUTTONS_HELP      = "Hides the fold buttons unless the mouse is over the gutter",
+        MAX_FOLD_LEVEL              = "Max fold level",
+        MAX_FOLD_LEVEL_HELP         = "Used to limit the number of nested folds to find and collapse when View -> Collapse All is called or Alt is held down when collapsing. Should improve performance for large files.";
```
This feels a little weird -- might be better for these to just be doc comments on the `definePreference()` calls for now... but not a big deal.  I think this would be ok to merge as-is.

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -273,6 +264,11 @@ define(function (require, exports, module) {
                 saveLineFolds(previous);
             }
+        } else {
+            if (current && current._codeMirror) {
```
Nit: can remove the `& current._codeMirror` here too

peterflynn replied:
Nm, this is taken care of in my PR

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -8,27 +8,35 @@
     prefs.definePreference("maxFoldLevel", "number", 2,
-                           {name: strings.MAX_FOLD_LEVEL, description: strings.MAX_FOLD_LEVEL_HELP});
+                           {name: MAX_FOLD_LEVEL, description: MAX_FOLD_LEVEL_HELP});
     prefs.definePreference("folds", "object", {});
```
This should be turned into `PreferencesManager.stateManager.definePreference(...`

peterflynn commented:
Looks like `minFoldSize` and `alwaysUseIndentFold` still require close & reopen to take effect.  Do you want to spin that off as a separate bug?

peterflynn commented:
Hmm, it also looks like folds are no longer remembered when you close & reopen a file, even with a completely clean slate of prefs.  Seems like something is broken with the view-state storage changes here...

thehogfather commented:
WRT `minFoldSize` and `alwaysUseIndentFold` lets file a separate bug with steps to reproduce. When I test it I get the effects immediately within the brackets.json file being edited (after saving).

peterflynn commented on src/extensions/default/CodeFolding/Prefs.js:
```diff
@@ -85,8 +93,9 @@ define(function (require, exports, module) {
     function setFolds(path, folds) {
-        store[path] = simplify(folds);
-        prefs.set(foldsKey, store);
+        var allFolds = PreferencesManager.getViewState(foldsKey);
```
You need `|| {}` here too -- that's why saving folds is broken currently.

peterflynn replied:
One other thought... rather than dumping folding state for all files together in one map, it might make sense to bucket it by project, like we do for other similar prefs like the scroll position & selection state of editors.

MainViewManager has an example of how to do that:
```
            context = { location : { scope: "user",
                                     layer: "project" } },
            state = PreferencesManager.getViewState(PREFS_NAME, context);
```

and

```
            context         = { location : { scope: "user",
                                         layer: "project",
                                         layerID: projectRoot.fullPath } },
        ...
        PreferencesManager.setViewState(PREFS_NAME, state, context);
```

abose replied:
@peterflynn  can you make the **|| {}** change in your PR?
Also issues#10869 added for tracking the bucketing by project.

abose replied:
brackets/issues/10870 opened to track Code Folding preference minFoldSize and alwaysUseIndentFold require close & reopen of file to take effect.

MarcelGerber replied:
I _think_ using the correct `PreferencesManager.stateManager.definePreference(...` will fix that, won't it?

peterflynn replied:
@MarcelGerber Yep, good point -- that should remove the need for `|| {}` both here & in `getFolds()`.

peterflynn commented on src/extensions/default/CodeFolding/main.js:
```diff
@@ -317,29 +416,12 @@ define(function (require, exports, module) {
             init();
```
Btw, why do we need to call `init()` again on project open?  Now that we're listening for preference changes at all times, it seems like this should be unnecessary?

peterflynn replied:
I removed this in my PR and things still seem to work ok

peterflynn commented:
Btw, it's still a ways off but it will be _interesting_ to think about what [having the same document visible in two split panes](https://trello.com/c/GezHZcCx/390-m-split-view-same-document) will mean for code folding.  Should collapsing a section in one pane automatically collapse it in the other?  Seems like that would feel weird, but otoh if each editor has its own expand/collapse state for the same document, how will that be persisted?  Which one "wins" if you close both views?  Etc.

I'll add a note to that Trello story so we can think about it more when the time comes.

abose commented:
Loos good now; Merging this PR.
@peterflynn brackets/pull/10859 can now be merged.

MarcelGerber commented:
@peterflynn It's the same issue for selections/cursors, having two states in one document is weird anyway.

peterflynn commented:
@abose Wait, why did you merge this already?  See my comments above -- https://github.com/adobe/brackets/pull/10850#issuecomment-91371932 and https://github.com/adobe/brackets/pull/10850#discussion_r28110161 -- code folding persistence is completely broken in master now because of this PR.  We should have waited before fixing that blocker before merging...

I'll file a high-priority bug to track it.
